go/enclave/rpc/rpc_utils.go

```diff
@@ -0,0 +1,135 @@
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+
+	"github.com/ethereum/go-ethereum/core/types"
+
+	"github.com/obscuronet/go-obscuro/go/common"
+
+	gethcommon "github.com/ethereum/go-ethereum/common"
+)
+
+const (
+	// CallFieldTo and CallFieldFrom and CallFieldData are the relevant fields in a Call request's params.
+	CallFieldTo   = "to"
+	CallFieldFrom = "from"
+	CallFieldData = "data"
+)
+
+// ExtractTxHash returns the transaction hash from the params of an eth_getTransactionReceipt request.
+func ExtractTxHash(getTxReceiptParams []byte) (gethcommon.Hash, error) {
+	var paramsJSONList []string
+	err := json.Unmarshal(getTxReceiptParams, &paramsJSONList)
+	if err != nil {
+		return gethcommon.Hash{}, fmt.Errorf("could not parse JSON params in eth_getTransactionReceipt "+
+			"request. JSON params are: %s. Cause: %w", string(getTxReceiptParams), err)
+	}
+	txHash := gethcommon.HexToHash(paramsJSONList[0]) // The only argument is the transaction hash.
+	return txHash, err
+}
+
+// ExtractTx returns the common.L2Tx from the params of an eth_sendRawTransaction request.
+func ExtractTx(sendRawTxParams []byte) (*common.L2Tx, error) {
+	// We need to extract the transaction hex from the JSON list encoding. We remove the leading `"[0x`, and the trailing `]"`.
+	txBinary := sendRawTxParams[4 : len(sendRawTxParams)-2]
+	txBytes := gethcommon.Hex2Bytes(string(txBinary))
+
+	tx := &common.L2Tx{}
+	err := tx.UnmarshalBinary(txBytes)
+	if err != nil {
+		return nil, fmt.Errorf("could not unmarshall transaction from bytes. Cause: %w", err)
+	}
+
+	return tx, nil
+}
+
+// GetViewingKeyAddressForTransaction returns the address whose viewing key should be used to encrypt the response,
+// given a transaction.
+func GetViewingKeyAddressForTransaction(tx *common.L2Tx) (gethcommon.Address, error) {
+	// TODO - Once the enclave's genesis.json is set, retrieve the signer type using `types.MakeSigner`.
+	signer := types.NewLondonSigner(tx.ChainId())
+	sender, err := signer.Sender(tx)
+	if err != nil {
+		return gethcommon.Address{}, fmt.Errorf("could not recover sender for transaction. Cause: %w", err)
+	}
+	return sender, nil
+}
+
+// GetViewingKeyAddressForBalanceRequest returns the address whose viewing key should be used to encrypt the response,
+// given the params of an eth_getBalance request.
+func GetViewingKeyAddressForBalanceRequest(balanceParams []byte) (gethcommon.Address, error) {
+	var paramsJSONMap []string
+	err := json.Unmarshal(balanceParams, &paramsJSONMap)
+	if err != nil {
+		return gethcommon.Address{}, fmt.Errorf("could not parse JSON params in eth_getBalance request. JSON "+
+			"params are: %s. Cause: %w", string(balanceParams), err)
+	}
+	// The first argument is the address, the second the block.
+	return gethcommon.HexToAddress(paramsJSONMap[0]), nil
+}
+
+// ExtractCallParamTo extracts and parses the `to` field of an eth_call request.
+func ExtractCallParamTo(callParams []byte) (gethcommon.Address, error) {
```

Comment:
Could we have a CallParams struct and then just a single `ExtractCallParams` method that returns all of them bundled together. Would clear up the duplication here both in code and in CPU work,

Reply:
I would do that, but on the WE side, we only need to extract one of them (the `from`), so I thought this was the best.

Reply:
Personally I'd rather extract the fields unnecessarily for the WE than duplicate the code and work in the enclave but don't feel strongly.